feat: implement open feature provider

duyhungtnn opened the pull request:
Part of https://github.com/bucketeer-io/bucketeer/issues/887

This PR introduces the Buckeeteer Open Feature provider

duyhungtnn committed:
feat: implement open feature provider

kakcy commented:
`.podspec` not found.
The [iOS SDK](https://github.com/bucketeer-io/ios-client-sdk) is also distributed with `CocoaPods`, so it would be a good idea to support it.

kakcy commented:
Are you planning to write a ReadMe?
It would be better to have a ReadMe with the same content as the [iOS SDK](https://github.com/bucketeer-io/ios-client-sdk/blob/main/README.md).

duyhungtnn commented:
@kakcy thanks for your suggestion. I will update a README file.

I am not planning to support CocoaPods because the OpenFeature iOS SDK does not officially support it. They only provide SPM package

kakcy commented:
Since `SwiftLint` is used in this project, I would like you to commit `.swiftlint.yml` to maintain the uniformity of the development environment.

duyhungtnn commented:
hi @cre8ivejp @kakcy 
I updated the [README](https://github.com/bucketeer-io/openfeature-swift-client-sdk/blob/feat/init-provider/README.md) and added `.swiftlint.yml`

kakcy commented on README.md:
```diff
@@ -1 +1,220 @@
+
+This is the official Swift OpenFeature provider for accessing your feature flags with [Bucketeer](https://bucketeer.io/).
+
+[Bucketeer](https://bucketeer.io) is an open-source platform created by [CyberAgent](https://www.cyberagent.co.jp/en) to help teams make better decisions, reduce deployment lead time and release risk through feature flags. Bucketeer offers advanced features like dark launches and staged rollouts that perform limited releases based on user attributes, devices, and other segments.
```
```suggestion
[Bucketeer](https://bucketeer.io) is an open-source platform created by [CyberAgent](https://www.cyberagent.co.jp/en/) to help teams make better decisions, reduce deployment lead time and release risk through feature flags. Bucketeer offers advanced features like dark launches and staged rollouts that perform limited releases based on user attributes, devices, and other segments.
```

duyhungtnn replied:
@kakcy did you set the API Key & API Domain ?

kakcy reviewed (commented):
The following tests are unstable and cause errors in E2E testing. 
`E2EProviderDestroyAndReinitializeTests.destroyAndReinitialize()`

```
Failing tests:
E2EProviderDestroyAndReinitializeTests.destroyAndReinitialize()
E2EProviderGetEvaluationTests.shouldUpdateTheUserAtributeWhenContextChange()
E2EProviderGetEvaluationTests.shouldUpdateTheUserAtributeWhenContextChange()
E2EProviderGetEvaluationTests.shouldNotUpdateTheUserAtribute()
E2EProviderGetEvaluationTests.shouldNotUpdateTheUserAtribute()
```
The tests that can cause errors vary depending on the timing, but the errors listed above are just a few.
If I remove the tests, the tests become stable.

Do you get the error, too?
Do you know what causes it?

duyhungtnn commented:
hi @kakcy , I didn't see that error before.  But now when I tried the e2e, it fail like you saw too.
It should be related to a race condition when the e2e test should only run in the `serial` mode.

duyhungtnn commented:
hi @kakcy its a problem with XCode 16 & Swift Testing.
XCode 16 will try to run Swift Testing in parallelization.
I was looking on how to set that using XCodeGen, but I didn’t found a way to setup that with XCodeGen in the current version.
It seems Swift Testing in XCode 16 is too new with XCodeGen and its not support yet.
I rewrite the E2e Test using XCTest.

https://github.com/bucketeer-io/openfeature-swift-client-sdk/pull/1/commits/a32824ac9e6295b42d26d22cb19fa4b187a050ae

kakcy commented:
@duyhungtnn 
Thank you for responding to various points of concern!
LGTM 👍

cre8ivejp commented on README.md:
````diff
@@ -1 +1,220 @@
+}
+```
+
+See our [documentation](https://docs.bucketeer.io/sdk/client-side/ios) for more SDK configuration..
+
+The evaluation context is the way for the client to specify contextual data that Bucketeer uses to evaluate the feature flags.
+
+The `targetingKey` should be the id of a user and should not be empty string.
+
+### Update the Evaluation Context
+
+If the user attributes change, you can update the evaluation context with the new attributes.
````
```suggestion
See our [documentation](https://docs.bucketeer.io/sdk/client-side/ios) for more SDK configuration..

The evaluation context allows the client to specify contextual data that Bucketeer uses to evaluate the feature flags.

The `targetingKey` is the user ID (Unique ID) and cannot be empty.

### Update the Evaluation Context

You can update the evaluation context with the new attributes if the user attributes change.
```

cre8ivejp commented on README.md:
```diff
@@ -1 +1,220 @@
+> [!WARNING]
+> Changing the `targetingKey` is not supported in the current implementation of the BucketeerProvider.
+
+To handle a user ID change, the BucketeerProvider must be remove and reinitialized.
```
```suggestion
To change the user ID, the BucketeerProvider must be removed and reinitialized.
```

cre8ivejp commented on README.md:
```diff
@@ -1 +1,220 @@
+
+### Initialize the provider
+
+Bucketeer provider needs to be created and then set in the global OpenFeatureAPI. 
```
```suggestion
Bucketeer provider needs to be created and then set in the global OpenFeatureAPI.
```

cre8ivejp committed:
ci: add release workflow

Signed-off-by: Alessandro Yuichi Okimoto <yuichijpn@gmail.com>

cre8ivejp committed:
docs: add changelog

Signed-off-by: Alessandro Yuichi Okimoto <yuichijpn@gmail.com>

cre8ivejp reviewed (approved):
Thank you for the awesome work!
